### ISSUE TITLE
Replace Hcaptcha by Cloudflare Turnstile

### ISSUE THREAD

**Issue by azam-aziz** [pentacent/keila#362]
Hi. I thought that it is a good option to use cloudflare turnstile as it is free and now big projects and services are using it which means it is reliable. What are your thoughts ?

**Comment by azamaziz989** [comment]
Hi. Recently i found out that Cloudflare turnstile only works for authorized and selected domains so that would not work for forms.

**Comment by Omikorin** [comment]
I'd like to recommend Proof-of-Work captchas like [mCaptcha](https://github.com/mCaptcha/mCaptcha) or [altcha](https://github.com/altcha-org/altcha). Dunno if they are compatible with keila's current approach.

**Comment by BloodyIron** [comment]
Yeah can we please have self-hosted captcha options added? mCaptcha, libreCaptcha, maybe some others, unsure which to prefer (maybe mCaptcha?). Having to rely on provided services when self-hosting Keila works against the intersts of those who want fully self-hosted systems.

**Comment by wmnnd** [comment]
I currently don't have plans to add alternative captcha providers to Keila, but I'm open to reviewing merging PRs for mCaptcha, altcha, or libreCaptcha (probably not Cloudflare though). It’s actually quite straightforward to add another Captcha provider, just take a look at `lib/keila_web/helpers/captcha/captcha.ex` to see how we're already supporting hCaptcha and Friendly Captcha. The other place where you have to make an adjustment is `runtime.exs`.

Actually, I just noticed there’s a bug in the captcha configuration that no one has discovered so far. I’ll fix it right away so that adding a new captcha provider should be really easy.

**Comment by wmnnd** [comment]
Just pushed a change that allows setting `CAPTCHA_SCRIPT_URL`, so now adding another provider that works similar to reCaptcha/hCaptcha and also allows for fully self-hosted URLs should be a piece of cake 🍰 😊

**Comment by BloodyIron** [comment]
> Just pushed a change that allows setting `CAPTCHA_SCRIPT_URL`, so now adding another provider that works similar to reCaptcha/hCaptcha and also allows for fully self-hosted URLs should be a piece of cake 🍰 😊

Does this include alts like mCaptcha/libreCaptcha?

**Comment by wmnnd** [comment]
The change makes it possible to add support for a new provider, provided they work similarly to reCaptcha (i.e. embed them with a single line of JS, a property called `data-sitekey` and then a similar POST callback with just a handful of lines. If the provider works a bit differently, it needs a few more lines 😊

**Comment by BloodyIron** [comment]
Cool, thanks!
